hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HStore.java

```diff
@@ -2360,149 +2364,15 @@ public void upsert(Iterable<Cell> cells, long readpoint, MemStoreSizing memstore
-      return HStore.this.updateStorefiles(storeFiles, snapshot.getId());
-    }
-
-    @Override
-    public long getOutputFileSize() {
-      return outputFileSize;
-    }
-
-    @Override
-    public List<Path> getCommittedFiles() {
-      return committedFiles;
-    }
-
-    /**
-     * Similar to commit, but called in secondary region replicas for replaying the
-     * flush cache from primary region. Adds the new files to the store, and drops the
-     * snapshot depending on dropMemstoreSnapshot argument.
-     * @param fileNames names of the flushed files
-     * @param dropMemstoreSnapshot whether to drop the prepared memstore snapshot
-     */
-    @Override
-    public void replayFlush(List<String> fileNames, boolean dropMemstoreSnapshot)
-        throws IOException {
-      List<HStoreFile> storeFiles = new ArrayList<>(fileNames.size());
-      for (String file : fileNames) {
-        // open the file as a store file (hfile link, etc)
-        StoreFileInfo storeFileInfo =
-          getRegionFileSystem().getStoreFileInfo(getColumnFamilyName(), file);
-        HStoreFile storeFile = createStoreFileAndReader(storeFileInfo);
-        storeFiles.add(storeFile);
-        HStore.this.storeSize.addAndGet(storeFile.getReader().length());
-        HStore.this.totalUncompressedBytes
-            .addAndGet(storeFile.getReader().getTotalUncompressedBytes());
-        if (LOG.isInfoEnabled()) {
-          LOG.info(this + " added " + storeFile + ", entries=" + storeFile.getReader().getEntries() +
-              ", sequenceid=" + storeFile.getReader().getSequenceID() + ", filesize="
-              + TraditionalBinaryPrefix.long2String(storeFile.getReader().length(), "", 1));
-        }
-      }
-
-      long snapshotId = -1; // -1 means do not drop
-      if (dropMemstoreSnapshot && snapshot != null) {
-        snapshotId = snapshot.getId();
-        snapshot.close();
-      }
-      HStore.this.updateStorefiles(storeFiles, snapshotId);
-    }
-
-    /**
-     * Abort the snapshot preparation. Drops the snapshot if any.
-     */
-    @Override
-    public void abort() throws IOException {
-      if (snapshot != null) {
-        //We need to close the snapshot when aborting, otherwise, the segment scanner
-        //won't be closed. If we are using MSLAB, the chunk referenced by those scanners
-        //can't be released, thus memory leak
-        snapshot.close();
-        HStore.this.updateStorefiles(Collections.emptyList(), snapshot.getId());
-      }
+  public StoreFlushContext createFlushContext(long cacheFlushId, FlushLifeCycleTracker tracker)
+      throws IOException {
+    Class<StoreFlushContext> flushContextClass = (Class<StoreFlushContext>)
+      conf.getClass(STORE_FLUSH_CONTEXT_CLASS_NAME, DefaultStoreFlushContext.class);
+    try {
+      return flushContextClass.getConstructor(HStore.class, Long.class, FlushLifeCycleTracker.class)
+        .newInstance(this, cacheFlushId, tracker);
```

Comment:
Do you think it would be cleaner to move this into an `init()` method rather than using reflection on the constructor?